internal/cmd/docker/push.go

```diff
@@ -96,11 +99,92 @@ func pushDockerImage(imageName, username, password string, timeout time.Duration
 		return nil
 	}
 
-	// Print the push output
-	_, err = io.Copy(os.Stdout, out)
-	if err != nil {
-		return fmt.Errorf("docker output: %v", err)
+	return logPushProgress(out)
+}
+
+func logPushProgress(reader io.ReadCloser) error {
+	var stepID string
+	bars := map[string]*progressbar.ProgressBar{}
+	decoder := json.NewDecoder(reader)
+	for {
+		// Decode the message from the Docker API output.
+		var msg dockermsg.JSONMessage
+		err := decoder.Decode(&msg)
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return fmt.Errorf("failed to decode JSON during Docker image push: %v", err)
+		}
+		if msg.Error != nil {
+			return fmt.Errorf("server error during Docker image push: %s", msg.Error.Message)
+		}
+
+		// Create a new progress spinner or bar when the Docker push ID changes.
+		// Each ID corresponds to a distinct step in the push process.
+		// Note: Outputs are in parallel; identical IDs indicate outputs from the same thread.
+		if stepID != msg.ID {
+			stepID = msg.ID
+			progress.Stop()
+
+			// Create a progress spinner for statuses that don't have progress details, like 'Preparing'.
+			if msg.Progress == nil || msg.Progress.Total == 0 {
+				progress.Show(msg.Status)
+				continue
+			}
+
+			// Init a progress bar for 'Pushing' status with total bytes and add it to bar group.
+			if _, ok := bars[msg.ID]; !ok {
+				bars[msg.ID] = createBar(msg.Progress.Total, fmt.Sprintf("%s %s", msg.Status, msg.ID))
+			}
+		}
+
+		// Update current progress based on msg.Progress.Total when in 'Pushing' status.
+		// Note: The Docker API may return a current value greater than total. To prevent breaking the progress bar,
+		// only update when the current is less than total.
```

Comment:
Does that really break the progress bar?

This works fine for me, even when the loop goes beyond the max total of the bar.
```go
func main() {
	bar := progressbar.Default(50)
	for i := 0; i < 100; i++ {
		_ = bar.Add(1)
		time.Sleep(40 * time.Millisecond)
	}
}
```

Reply:
But once you finish it, it'd return this error.
https://github.com/saucelabs/saucectl/pull/913#issuecomment-2134045121

Reply:
I don't use `Add`. `Set64` can break it.

```
func main() {
	bar := progressbar.Default(50)
	bar.Set64(3000)
	err := bar.Finish()
	fmt.Println("err: ", err)
}
```

Reply:
https://github.com/schollz/progressbar/blob/v3.14.3/progressbar.go#L578

Reply:
😄 100% break
```
func main() {
	bar := progressbar.Default(50)
	for i := 0; i < 100; i++ {
		err := bar.Add(3)
		if err != nil {
			fmt.Println("err: ", err)
			break
		}
		time.Sleep(40 * time.Millisecond)
	}
}
```